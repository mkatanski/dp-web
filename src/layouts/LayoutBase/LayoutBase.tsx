import React, { Suspense, ComponentType } from "react";
import { Route, RouteProps, Switch } from "react-router-dom";
import { LoadingScreen } from "components/fallbacks/LoadingScreen";

const NotFoundPage = React.lazy(() => import("pages/errors/NotFoundPage"));

export interface LayoutBaseProps extends RouteProps {
  /**
   * Component rendered by fallback route. By default its 'views/errors/NotFoundView'
   */
  fallbackRouteComponent?: ComponentType;
  /**
   * If true append default route if react-router switch won't match current path
   */
  useFallbackRoute?: boolean;
}

const LayoutBase: React.FC<LayoutBaseProps> = ({
  children,
  fallbackRouteComponent,
  useFallbackRoute = true,
  ...rest
}: LayoutBaseProps) => (
  <Route
    {...rest}
    render={() => (
      <Suspense fallback={<LoadingScreen />}>
        <Switch>
          {children}
          {useFallbackRoute && (
            <Route component={fallbackRouteComponent || NotFoundPage} />
          )}
        </Switch>
      </Suspense>
    )}
  />
);

export default LayoutBase;
