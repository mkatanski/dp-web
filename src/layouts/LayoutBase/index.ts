import LayoutBase from "./LayoutBase";
export * from "./LayoutBase";

export default LayoutBase;
