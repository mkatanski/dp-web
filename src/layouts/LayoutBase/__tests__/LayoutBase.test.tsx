import React from "react";
import TestRenderer from "react-test-renderer";

import LayoutBase from "../LayoutBase";
import { wait } from "@testing-library/react";

jest.mock(
  "pages/errors/NotFoundPage",
  () => ({ children, ...rest }: React.PropsWithChildren<{}>) => (
    <div id="NotFoundPage" {...rest}>
      {children}
    </div>
  )
);

jest.mock("components/fallbacks/LoadingScreen", () => ({
  LoadingScreen: jest.fn(() => {
    return <div id="LoadingScreen" />;
  })
}));

describe("layouts", () => {
  describe("LayoutBase", () => {
    it("should render LayoutBase component w/o fallback", async () => {
      const tree = TestRenderer.create(
        <LayoutBase useFallbackRoute={false}>children</LayoutBase>
      );
      expect(tree).toMatchSnapshot();
    });

    it("should render LayoutBase with given fallback component", () => {
      const tree = TestRenderer.create(
        <LayoutBase fallbackRouteComponent={() => <div>Fallback</div>}>
          children
        </LayoutBase>
      );
      expect(tree).toMatchSnapshot();
    });

    it("should render LayoutBase with default fallback component", async () => {
      const tree = TestRenderer.create(<LayoutBase>children</LayoutBase>);
      await wait();

      expect(tree).toMatchSnapshot();
    });

    it("should render a spinner as a fallback", () => {
      const Test = () => {
        throw new Promise(r => {
          r();
        });
      };
      const tree = TestRenderer.create(
        <LayoutBase>
          <Test />
        </LayoutBase>
      );
      expect(tree).toMatchSnapshot();
    });
  });
});
