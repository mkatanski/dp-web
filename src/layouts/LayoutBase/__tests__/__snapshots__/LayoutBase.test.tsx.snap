// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layouts LayoutBase should render LayoutBase component w/o fallback 1`] = `
<div
  id="ReactRouter.Route"
>
  <div
    id="ReactRouter.Switch"
  >
    children
  </div>
</div>
`;

exports[`layouts LayoutBase should render LayoutBase with default fallback component 1`] = `
<div
  id="ReactRouter.Route"
>
  <div
    id="ReactRouter.Switch"
  >
    children
    <div
      id="ReactRouter.Route"
    >
      <div
        id="NotFoundPage"
      />
    </div>
  </div>
</div>
`;

exports[`layouts LayoutBase should render LayoutBase with given fallback component 1`] = `
<div
  id="ReactRouter.Route"
>
  <div
    id="ReactRouter.Switch"
  >
    children
    <div
      id="ReactRouter.Route"
    >
      <div>
        Fallback
      </div>
    </div>
  </div>
</div>
`;

exports[`layouts LayoutBase should render a spinner as a fallback 1`] = `
<div
  id="ReactRouter.Route"
>
  <div
    id="LoadingScreen"
  />
</div>
`;
